common/models/checkpoint.js

```diff
@@ -27,43 +28,59 @@ module.exports = function(Checkpoint) {
    * Get the current checkpoint id
    * @callback {Function} callback
    * @param {Error} err
-   * @param {Number} checkpointId The current checkpoint id
+   * @param {Number} checkpoint The current checkpoint seq
    */
-
   Checkpoint.current = function(cb) {
     var Checkpoint = this;
-    this.find({
-      limit: 1,
-      order: 'seq DESC'
-    }, function(err, checkpoints) {
-      if (err) return cb(err);
-      var checkpoint = checkpoints[0];
-      if (checkpoint) {
-        cb(null, checkpoint.seq);
-      } else {
-        Checkpoint.create({ seq: 1 }, function(err, checkpoint) {
-          if (err) return cb(err);
-          cb(null, checkpoint.seq);
-        });
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (cp) {
+          cb(null, cp.seq);
+        } else {
+          Checkpoint.init(function(err, data) {
+            cb(err, data.seq);
+          });
+        }
       }
-    });
+    ],
+    cb);
   };
 
-  Checkpoint.observe('before save', function(ctx, next) {
-    if (!ctx.instance) {
-      // Example: Checkpoint.updateAll() and Checkpoint.updateOrCreate()
-      return next(new Error('Checkpoint does not support partial updates.'));
-    }
+  /**
+   * Initialize the checkpoint
+   * @callback {Function} callback
+   * @param {Error} err
+   * @param {Object} checkpoint The current checkpoint
+   */
+  Checkpoint.init = function(cb) {
+    this.create({seq: 1}, cb);
+  };
 
-    var model = ctx.instance;
-    if (!model.getId() && model.seq === undefined) {
-      model.constructor.current(function(err, seq) {
-        if (err) return next(err);
-        model.seq = seq + 1;
-        next();
-      });
-    } else {
-      next();
-    }
-  });
+  /**
+   * Increase the current checkpoint If it already exists otherwise initialize it
+   * @callback {Function} callback
+   * @param {Error} err
+   * @param {Object} checkpoint The current checkpoint
+   */
+  Checkpoint.bump = function(cb) {
+    var Checkpoint = this;
+
+    async.waterfall([
+      function(cb) {
+        Checkpoint.findOne(cb);
+      },
+      function(cp, cb) {
+        if (!cp) {
+          Checkpoint.init(cb);
+        } else {
+          cp.seq++;
+          cp.save(cb);
+        }
+      }
+    ], cb);
```

Comment:
Hmm, this will be tricky to implement in an atomic fashion. I wish we had a method like `updateOrCreate(filter, { seq: { $inc: 1 } }, { seq: 1})`.

I think this should be good enough for now:

```js
// todo - this is shared with .current(), refactor it to a shared private function, e.g. _getSingleton()
Checkpoint.findOrCreate({}, { seq: 1 }, function(err, cp) {
  // todo - handle errors

  var originalSeq = cp.seq;
  cp.seq++;

  // Update the checkpoint but only if it was not changed under our hands
  Checkpoint.updateAll({ id: cp.id, seq: originalSeq }, { seq: cp.seq }, function(err, info) {
    // todo handle errors

    // possible outcomes - please preserve this info in a code comment
    // 1) seq was updated to seq+1 - exactly what we wanted!
    // 2) somebody else already updated seq to seq+1 and our call was a no-op. 
    //   That should be ok, checkpoints are time based, so we reuse the one created just now
   //  3) seq was bumped more than once, so we will be using a value that is behind the latest seq.
   //    I am not entirely sure if this is ok, but since it wasn't handled by the current implementation either,
   //    I think we can keep it this way.
    cb(null, cp);
}
```